infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationDiscoveryService.java

```diff
@@ -405,29 +406,52 @@ private void registerToLogReplicationLock() {
     }
 
     /**
-     * This method is only called on the leader node and it triggers the start of log replication
+     * This method is only called on the leader node and it triggers the start of log replication and on topology change
      * <p>
-     * Depending on the role of the cluster to which this leader node belongs, it will start
-     * as source (sender/producer) or sink (receiver).
+     * Depending on the role of the cluster to which this leader node belongs (Source, Sink or both), it will start
+     * respective components
      */
     private void onLeadershipAcquire() {
-        // TODO [V2]: a cluster can be Source and/or Sink for different replication models. The support for this is
-        //  coming in the connectionModel PR
-        if (isSource()) {
-            log.info("Start as Source (sender/replicator)");
-            sessionManager.startReplication();
-            lockAcquireSample = recordLockAcquire();
-            processCountOnLockAcquire();
-        } else if (isSink()) {
-            // TODO: Shama: in connectionModel PR, log session info when cluster is a sink
-            log.info("Start as Sink (receiver)");
-            // Sink Site : the LogReplicationServer (server handler) will
-            // reset the LogReplicationSinkManager on acquiring leadership
-            interClusterServerNode.setLeadership(true);
-            lockAcquireSample = recordLockAcquire();
-            processCountOnLockAcquire();
-        } else {
+        if (!isSource() && !isSink()) {
             log.error("Log Replication not started on this cluster. Remote source/sink not found");
+            return;
+        }
+
+        setupLogReplicationServer();
+
+        sessionManager.startConnectionReceivingRouters(isSource(), isSink(), serverMap, interClusterServerNode);
+        sessionManager.connectToRemoteClusters(serverMap);
+
+        lockAcquireSample = recordLockAcquire();
```

Comment:
Plus a comment on this lock. Why it is needed?

Reply:
This one here is for metrics. This line and the one below it record metrics around the lock. 
Have added a comment saying that.

But the lock in this file determines which node in a cluster is the leader. All the nodes compete for the lock and one succeeds.